wgengine/netstack/netstack.go

```diff
@@ -302,6 +352,74 @@ func (ns *Impl) wrapProtoHandler(h func(stack.TransportEndpointID, stack.PacketB
 	}
 }
 
+// wrapTCPProtocolHandler wraps the protocol handler we pass to netstack for TCP.
+func (ns *Impl) wrapTCPProtocolHandler(h protocolHandlerFunc) protocolHandlerFunc {
+	return func(tei stack.TransportEndpointID, pb stack.PacketBufferPtr) (handled bool) {
+		localIP, ok := netip.AddrFromSlice(tei.LocalAddress.AsSlice())
+		if !ok {
+			ns.logf("netstack: could not parse local address for incoming connection")
+			return false
+		}
+		localIP = localIP.Unmap()
+
+		remoteIP, ok := netip.AddrFromSlice(tei.RemoteAddress.AsSlice())
+		if !ok {
+			ns.logf("netstack: could not parse remote address for incoming connection")
+			return false
+		}
+
+		// If we have too many pending connections for this client, abort
+		// early and don't open a new one.
+		//
+		// NOTE: the counter is decremented in
+		// decrementPendingTCPForward, called from the acceptTCP
+		// function, below.
+		ns.mu.Lock()
+		pending := ns.connsPendingByClient[remoteIP]
+		tooManyPending := pending > maxInFlightConnectionAttemptsPerClient()
+		if !tooManyPending {
+			ns.connsPendingByClient[remoteIP]++
+		}
+		ns.mu.Unlock()
+		if debugNetstack() {
+			ns.logf("[v2] netstack: pending connections for client %v: %d", remoteIP, pending)
+		}
+		if tooManyPending {
+			return false // unhandled
```

Comment:
I think we'd want to log these by default to make it easier to debug connection establishment issues.

```suggestion
			ns.logf("netstack: ignoring a new TCP connection from %v to %v because the client already has %d pending connections", localIP, remoteIP, pending)
			return false // unhandled
```